File: ada-boost-implementations/code-java/lib/src/main/java/vkmbox/classifier/base/DecisionStumpContinuous.java
package vkmbox.classifier.base;

import org.nd4j.common.util.ArrayUtil;
import org.nd4j.linalg.api.ndarray.INDArray;

/**
 *  DecisionStump-Continuous classifier.
 *  Classifies: if value > threshold then sign else -1 * sign, 
 *  where sign = {-1,+1}
 */
public class DecisionStumpContinuous {
    
    private final int featureNumber;
    private final int sign;
    private final double threshold;
    
    public DecisionStumpContinuous(int featureNumber, int sign, double threshold) {
        this.featureNumber = featureNumber;
        this.sign = sign;
        this.threshold = threshold;
    }

    public int[] classify(INDArray dataX) {
        return getClassification(dataX, featureNumber, threshold, sign);
    }

    public static int[] getClassification(INDArray dataX, int featureNumber
            , double threshold, int sign) {
        double[] featureX = ArrayUtil.flattenDoubleArray(dataX.getColumn(featureNumber));
        int[] result = new int[featureX.length];
        for (int idx = 0; idx < featureX.length; idx++) {
            result[idx] = featureX[idx] > threshold ? sign: -1*sign;
        }
        return result; //[sign if value > threshold else -sign for value in X[:, feature_number]]
    }

    public static double getError(double[] featureX, int[] dataY, double[] weights
            , int sign, double threshold) {
        double error = 0.0;
        for (int idx = 0; idx < featureX.length; idx++) {
            int value = featureX[idx] > threshold ? sign: -1*sign;
            error += Math.abs((value-dataY[idx])/2)*weights[idx];
        }
        return error;
    }

    @Override
    public String toString() {
        return String.format("feature_number: %d, sign: %d, threshold: %f"
                , featureNumber, sign, threshold);
    }
}

File: ada-boost-implementations/code-python/decision_stump_continuous.py
class DecisionStumpContinuous:
    """
    DecisionStump-Continuous classifier.
    Classifies: if value > threshold then sign else -1 * sign, 
    where sign = {-1,+1}
    """
    
    def __init__(self, feature_number, sign, threshold):
        self.feature_number = feature_number
        self.threshold = threshold
        self.sign = sign

    def classify(self, X):
        return self.get_classification(X, self.feature_number, self.threshold, self.sign)

    @staticmethod
    def get_classification(X, feature_number, threshold, sign):
        return [DecisionStumpContinuous.get_classification_scalar(value, threshold, sign) for value in X[:, feature_number]]

    @staticmethod
    def get_classification_scalar(x_scalar, threshold, sign):
        return sign if x_scalar > threshold else -sign

    #(X, y, feature_number, d_t, sign, threshold)
    @staticmethod
    def get_error(X, y, feature_number, d_t, sign, threshold):
        error = 0
        for num in range(len(y)): #arg, fn, weight in zip(feature, y, d_t):
            arg, fn, weight = X[num, feature_number], y[num], d_t[num]
            val = sign if arg > threshold else -1*sign
            error += abs((val-fn)/2)*weight
            
        return error

    def str(self):
        return "feature_number: {}, sign: {}, threshold: {}".format(self.feature_number, self.sign, self.threshold)
